sigstore/_internal/rekor/checkpoint.py

```diff
@@ -0,0 +1,229 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Rekor Checkpoint machinery.
+"""
+
+from __future__ import annotations
+
+import base64
+import re
+import struct
+from dataclasses import dataclass
+from typing import List
+
+from pydantic import BaseModel, Field, StrictStr
+
+from sigstore._internal.keyring import KeyringSignatureError
+from sigstore._internal.rekor.client import RekorClient
+from sigstore._utils import KeyID
+from sigstore.transparency import LogEntry
+
+
+@dataclass(frozen=True)
+class RekorSignature:
+    """
+    Represents a `RekorSignature` containing:
+    - the name of the signature, e.g. "rekor.sigstage.dev"
+    - the signature hash
+    - the base64 signature
+    """
+
+    name: str
+    sig_hash: bytes
+    signature: bytes
+
+
+class CheckpointError(Exception):
+    """Raised during LogCheckpoint parsing or verification."""
+
+
+class LogCheckpoint(BaseModel):
+    """
+    Represents a Rekor `LogCheckpoint` containing:
+    - an origin, e.g. "rekor.sigstage.dev - 8050909264565447525"
+    - the size of the log,
+    - the hash of the log,
+    - and any ancillary contants, e.g. "Timestamp: 1679349379012118479"
+    see: https://github.com/transparency-dev/formats/blob/main/log/README.md
+    """
+
+    origin: StrictStr
+    log_size: int
+    log_hash: StrictStr
+    other_content: List[str]
+
+    @classmethod
+    def from_text(cls, text: str) -> LogCheckpoint:
+        """
+        Serialize from the text header ("note") of a SignedNote.
+        """
+
+        lines = text.strip().split("\n")
+        if len(lines) < 4:
+            raise CheckpointError("Malformed LogCheckpoint: too few items in header!")
+
+        origin = lines[0]
+        if len(origin) == 0:
+            raise CheckpointError("Malformed LogCheckpoint: empty origin!")
+
+        log_size = int(lines[1])
+        root_hash = base64.b64decode(lines[2]).hex()
+
+        return LogCheckpoint(
+            origin=origin,
+            log_size=log_size,
+            log_hash=root_hash,
+            other_content=lines[3:],
+        )
+
+    @classmethod
+    def to_text(self) -> str:
+        """
+        Serialize a `LogCheckpoint` into text format.
+        See class definition for a prose description of the format.
+        """
+        return "\n".join(
+            [
+                self.origin,
+                str(self.log_size),
+                self.log_hash,
+            ]
+            + self.other_content
+        )
+
+
+@dataclass(frozen=True)
+class SignedNote:
+    """
+    Represents a signed `Note` containing a note and its corresponding list of signatures.
```

Comment:
There's no `Note` class, so maybe just:

```suggestion
    Represents a "signed note" containing a note and its corresponding list of signatures.
```